infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/server/mapping/SinglePortHostnameBuilder.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.docker.server.mapping;
+
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.StringJoiner;
+import java.util.regex.Pattern;
+
+/**
+ * Produces host names in form:
+ * [serverName].[machineName].[workspaceId].<external_or_internal_address>.<wildcardNipDomain>
+ *
+ * @author Max Shaposhnik (mshaposh@redhat.com)
+ */
+public class SinglePortHostnameBuilder {
+
+  private final String externalAddress;
+  private final String internalAddress;
+  private final String wildcardHost;
+
+  /**
+   * hostname labels may contain only the ASCII letters 'a' through 'z' (in a case-insensitive
+   * manner), the digits '0' through '9', and the hyphen ('-').
+   */
+  private final Pattern pattern = Pattern.compile("[^a-zA-Z0-9\\-]");
+
+  public SinglePortHostnameBuilder(
+      String externalAddress, String internalAddress, String wildcardHost) {
+    this.externalAddress = externalAddress;
+    this.internalAddress = internalAddress;
+    this.wildcardHost = wildcardHost;
+  }
+
+  /**
+   * Constructs hostname from given params.
+   *
+   * @param serverName optional server name
+   * @param machineName optional machine name
+   * @param workspaceID optional workspace ID
+   * @return composite hostname
+   */
+  public String build(String serverName, String machineName, String workspaceID) {
+    StringJoiner joiner = new StringJoiner(".");
+    if (serverName != null) {
+      joiner.add(normalize(serverName));
+    }
+    if (machineName != null) {
```

Comment:
Can you describe in javadocs the case when some component is null?

Reply:
yes